wrappers/vroom.rb

```diff
@@ -29,175 +29,329 @@ def initialize(hash = {})
-              i += 1
-            end
-            pos_rest += i
-          end
-          activities.insert(pos_rest, rest_id: rest.id)
+      cost = (result['summary']['cost'])
+      routes = result['routes'].map.with_index{ |route, index|
+        @previous = nil
+        vehicle = vrp.vehicles[index]
+        cost += vehicle.cost_fixed if route['steps'].size.positive?
+        activities = route['steps'].map{ |step|
+          read_step(vrp, vehicle, step)
+        }.compact
+        {
+          vehicle_id: vehicle.id,
+          original_vehicle_id: vehicle.original_id,
+          activities: activities,
+          start_time: activities.first[:begin_time],
+          end_time: activities.last[:begin_time] + (activities.last[:duration] || 0),
         }
-      else
-        rests.each{ |rest| activities.insert(vehicle_have_end ? -2 : -1, rest_id: rest.id) }
-      end
+      }
+
+      unassigneds = result['unassigned'].map{ |step| read_unassigned(vrp, step) }
+
+      log 'Solution cost: ' + cost.to_s + ' & unassigned: ' + unassigneds.size.to_s, level: :info
 
       {
         cost: cost,
-        costs: Models::Costs.new({}), # TODO: fulfill with solution costs
+        cost_details: Models::CostDetails.new({}), # TODO: fulfill with solution costs
         solvers: ['vroom'],
         elapsed: elapsed_time, # ms
 #        total_travel_distance: 0,
 #        total_travel_time: 0,
 #        total_waiting_time: 0,
 #        start_time: 0,
 #        end_time: 0,
-        routes: [{
-          start_time: 0,
-          end_time: activities.collect{ |a| a[:travel_time].to_f + (a[:detail] ? a[:detail][:duration].to_f : 0) }.reduce(&:+),
-          vehicle_id: vehicle.id,
-          original_vehicle_id: vehicle.original_id,
-          activities: activities
-        }],
-        unassigned: []
+        routes: routes,
+        unassigned: unassigneds
       }
     end
 
     private
 
-    def run_vroom(vehicles, services, points, matrices, dimensions, prefer_short, job)
+    def rest_equivalence(vrp)
+      rest_index = 0
+      @rest_hash = {}
+      vrp.vehicles.each{ |vehicle|
+        vehicle.rests.each{ |rest|
+          @rest_hash["#{vehicle.id}_#{rest.id}"] = {
+            index: rest_index,
+            vehicle: vehicle.id,
+            rest: rest
+          }
+          rest_index += 1
+        }
+      }
+      @rest_hash
+    end
+
+    def read_step(vrp, vehicle, step)
+      case step['type']
+      when 'start'
+        read_depot(vehicle, step)
+      when 'end'
+        read_depot(vehicle, step)
+      when 'job'
+        read_job(vrp, vehicle, step)
+      when 'break'
+        read_break(step)
+      when 'pickup'
+        read_shipment(vrp, vehicle, step)
+      when 'delivery'
+        read_shipment(vrp, vehicle, step)
+      else
+        raise
```

Comment:
We can merge the same items and order the checks so that it is more efficient (it is better if we check start and end last so that we that the condition is not executed for all services for example)

Also maybe an explanation for the raise

```suggestion
      case step['type']
      when 'job'
        read_job(vrp, vehicle, step)
      when 'break'
        read_break(step)
      when 'pickup', 'delivery'
        read_shipment(vrp, vehicle, step)
      when 'start', 'end'
        read_depot(vehicle, step)
      else
        raise 'Unimplemented stop type in wrappers/vroom.rb' 
```

Reply:
Does the performances of case depends on the order of the clauses ?

ok to merge the same items :+1:

Reply:
On C++ it does, on Ruby I am not 100% sure it depends on the underlying implementation. It cannot hurt I imagine.